### PR TITLE
repository: dedupe logic

### PR DESCRIPTION
**Pull request by abn**
This change unifies shared logic across repository implementations.

```

                                                              ┌───────────────────┐
                                                 ┌───────────►│InstalledRepository│
                                                 │            └───────────────────┘
                                                 │
                   ┌────────────────┐      ┌─────┴──────┐
                   │CachedRepository│◄─────┤ Repository │
                   └───────┬────────┘      └─────┬──────┘
                           │                     │
                           │                     │            ┌────┐
                           │                     └───────────►│Pool│
                           │                                  └────┘
                    ┌──────▼───────┐
         ┌──────────┤HTTPRepository│
         │          └──────┬───────┘
         │                 │
         │                 │
         │                 │
 ┌───────▼────────┐ ┌──────▼───────┐
 │LegacyRepository│ │PypiRepository│
 └────────────────┘ └──────────────┘
         ▲
         │
         │
         │
         │             ┌────────┐
         └─────────────┤HTMLPage│
            used by    └────────┘
                           ▲
                           │
                      ┌────┴─────┐
                      │LinkSource│
                      └──────────┘

```

## Testing
Examples clone poetry source from git, 
### Using pipx
```sh
pipx install --suffix=@3868 'poetry @ git+https://github.com/python-poetry/poetry.git@refs/pull/3868/head'
poetry@3868 install
```

### Using a container (podman | docker)
```sh
podman run --rm -i --entrypoint bash python:3.8 <<EOF
set -xe
python -m pip install -q git+https://github.com/python-poetry/poetry.git@refs/pull/3868/head
install -d foobar
pushd foobar
cat > pyproject.toml <<TOML
# add your pyproject.toml here
TOML
poetry install
EOF
```

### CODE REVIEW

File: src/poetry/repositories/repository.py [hunk 1]
```diff
@@ -2,18 +2,17 @@
...
 if TYPE_CHECKING:
     from poetry.core.packages.dependency import Dependency
     from poetry.core.packages.package import Package
-    from poetry.core.packages.utils.link import Link
 
 
 class Repository(BaseRepository):
     def __init__(self, packages: list[Package] = None, name: str = None) -> None:
-        super().__init__()
+        super().__init__(name)
```
Comment (radoering) [inline src/poetry/repositories/repository.py:15]:
```suggestion
        super().__init__(name, packages)
```

Then the rest of `__init__` can be dropped.

**Comment by tgolsson** [comment]
Heyo @abn! 

We chatted ~two weeks ago on Discord re some issues I had with secondary sources, and you directed me here! Since this PR would make all my fixes pointless (you've removed a lot of the affected code) I wanted to wait until after this PR, but I'm curious when it'll get merged.

**Comment by abn** [comment]
@tgolsson this is just waiting on a final review.

File: src/poetry/repositories/http.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+    def _download(self, url: str, dest: str) -> None:
+        return download_file(url, dest, session=self.session)
+
+    def _get_info_from_wheel(self, url: str) -> PackageInfo:
+        from poetry.inspection.info import PackageInfo
+
+        wheel_name = urllib.parse.urlparse(url).path.rsplit("/")[-1]
+        self._log(f"Downloading wheel: {wheel_name}", level="debug")
+
+        filename = os.path.basename(urllib.parse.urlparse(url).path.rsplit("/")[-1])
```
Comment (radoering) [inline src/poetry/repositories/http.py:112]:
```suggestion
        filename = os.path.basename(wheel_name)
```

File: src/poetry/repositories/http.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    def _get_info_from_sdist(self, url: str) -> PackageInfo:
+        from poetry.inspection.info import PackageInfo
+
+        sdist_name = urllib.parse.urlparse(url).path
+        sdist_name_log = sdist_name.rsplit("/")[-1]
+
+        self._log(
+            f"Downloading sdist: {sdist_name_log}",
+            level="debug",
+        )
```
Comment (radoering) [inline src/poetry/repositories/http.py:129]:
```suggestion
        self._log(f"Downloading sdist: {sdist_name_log}", level="debug")
```

File: src/poetry/repositories/link_sources/html.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+with warnings.catch_warnings():
+    warnings.simplefilter("ignore")
+    import html5lib
+
+
+class HTMLPage(LinkSource):
+    def __init__(self, url: str, content: str) -> None:
+        super().__init__(url=url)
+
+        self._content = content
```
Comment (radoering) [inline src/poetry/repositories/link_sources/html.py:23]:
```suggestion
```

File: src/poetry/repositories/link_sources/base.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            if pkg.name == name and pkg.version == version:
+                yield link
+
+    _clean_re = re.compile(r"[^a-z0-9$&+,/:;=?@.#%_\\|-]", re.I)
+
+    def clean_link(self, url: str) -> str:
+        """Makes sure a link is fully encoded.  That is, if a ' ' shows up in
+        the link, it will be rewritten to %20 (while not over-quoting
+        % or other characters)."""
+        return self._clean_re.sub(lambda match: f"%{ord(match.group(0)):02x}", url)
```
Comment (radoering) [inline src/poetry/repositories/link_sources/base.py:99]:
```suggestion
        return self.CLEAN_REGEX.sub(lambda match: f"%{ord(match.group(0)):02x}", url)
```

File: src/poetry/repositories/link_sources/base.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        name = canonicalize_name(name)
+
+        for link in self.links:
+            pkg = self.link_package_data(link)
+
+            if pkg.name == name and pkg.version == version:
+                yield link
+
+    _clean_re = re.compile(r"[^a-z0-9$&+,/:;=?@.#%_\\|-]", re.I)
+
```
Comment (radoering) [inline src/poetry/repositories/link_sources/base.py:94]:
```suggestion
```

File: src/poetry/repositories/link_sources/base.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        return Package(name, version, source_url=link.url)
+
+    def links_for_version(self, name: str, version: Version) -> Iterator[Link]:
+        name = canonicalize_name(name)
+
+        for link in self.links:
+            pkg = self.link_package_data(link)
+
+            if pkg.name == name and pkg.version == version:
```
Comment (radoering) [inline src/poetry/repositories/link_sources/base.py:90]:
```suggestion
            if pkg.name == name and pkg.version and pkg.version == version:
```

**Review by radoering — CHANGES_REQUESTED** [review]
Some general thoughts:

The methods `_links_to_data()`, `_get_response()`, `_get_page()` are defined in `HttpRepository` but are only used in `LegacyRepository` not `PypiRepository`. Is there a reason not to move them to `LegacyRepository`?

Considering the inheritance hierarchy:
```
BaseRepository
- Repository
  - InstalledRepository
  - TestRepository
- Pool
- CachedRepository
  - ...
```
Is `Repository` still required? I think its `__init__` and `find_packages` can be moved to `InstalledRepository`. `TestRepository` overrides `find_packages` anyway so it can inherit from `BaseRepository` directly. If I miss something and `Repository` is required, it should at least be renamed.

**Comment by abn** [comment]
I think unifiying `BaseRepository` and `Repository` makes sense. The original rationale was to keep the changes outside of `poetry.repositories` minimal. In retrospect that does not really matter.

As for inheritence, this is what it will look like eventually. Note the addition of find links related repositories, this is one of the reasons why some of the methods only used by legacy is in HTTP, as it conceptualyl makes sense for all HTTP backed repositories.

```

                                                              ┌───────────────────┐
                                                 ┌───────────►│InstalledRepository│
                                                 │            └───────────────────┘
                                                 │
                   ┌────────────────┐      ┌─────┴──────┐
                   │CachedRepository│◄─────┤ Repository ├──────────────────┐
                   └───────┬────────┘      └─────┬──────┘                  │
                           │                     │                         │
                           │                     │            ┌────┐       │
                           │                     └───────────►│Pool│       │
                           │                                  └────┘       │
                    ┌──────▼───────┐                                       │
         ┌──────────┤HTTPRepository├──────────────┐                        │
         │          └──────┬───────┘              │                        │
         │                 │                      │                        │
         │                 │                      │                        │
         │                 │                      │                        │
 ┌───────▼────────┐ ┌──────▼───────┐ ┌────────────▼──────────┐             │
 │LegacyRepository│ │PypiRepository│ │FindLinksHTMLRepository│             │
 └────────────────┘ └──────────────┘ └───────────────────────┘             │
         ▲                                 ▲                               │
         │                                 │                               │
         │                                 │                               ▼
         │                                 │          ┌────────────────────────────┐
         │             ┌────────┐          │          │FindLinksDirectoryRepository│
         └─────────────┤HTMLPage├──────────┘          └────────▲───────────────────┘
            used by    └────────┘   used by                    │
                           ▲                                   │ used by
                           │                                   │
                      ┌────┴─────┐                    ┌────────┴──────────┐
                      │LinkSource├───────────────────►│DirectoryLinkSource│
                      └──────────┘                    └───────────────────┘

```

**Review by radoering — APPROVED** [review]

**Pull request merged**